nvflare/app_common/executors/statistics/statistics_executor.py

```diff
@@ -0,0 +1,316 @@
+
+from nvflare.apis.dxo import DXO, DataKind
+from nvflare.apis.event_type import EventType
+from nvflare.apis.executor import Executor
+from nvflare.apis.fl_constant import ReservedKey, ReturnCode
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable, make_reply
+from nvflare.apis.signal import Signal
+from nvflare.app_common.abstract.statistics_spec import Statistics
+from nvflare.app_common.app_constant import StatisticsConstants as StC
+from nvflare.app_common.statistics.metic_config import MetricConfig
+from nvflare.app_common.statistics.numeric_stats import filter_numeric_features
+from nvflare.app_common.statistics.stats_def import Feature, Histogram
+from nvflare.app_common.validation_exception import ValidationException
+
+
+class StatisticsExecutor(Executor):
+    def __init__(
+        self,
+        generator_id: str,
+        min_count: int,
+        min_random: float,
+        max_random: float,
+    ):
+        super().__init__()
+        self.generator_id = generator_id
+        self.min_count = min_count
+        self.stats_generator: Optional[Statistics] = None
+        self.max_random = max_random
+        self.min_random = min_random
+
+    def metric_functions(self) -> dict:
+        return {
+            StC.STATS_COUNT: self.get_count,
+            StC.STATS_SUM: self.get_sum,
+            StC.STATS_MEAN: self.get_mean,
+            StC.STATS_STDDEV: self.get_stddev,
+            StC.STATS_VAR: self.get_variance_with_mean,
+            StC.STATS_HISTOGRAM: self.get_histogram,
+            StC.STATS_MAX: self.get_max_value,
+            StC.STATS_MIN: self.get_min_value,
+        }
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.START_RUN:
+            self.initialize(fl_ctx)
+        elif event_type == EventType.END_RUN:
+            self.finalize(fl_ctx)
+
+    def initialize(self, fl_ctx: FLContext):
+        try:
+            engine = fl_ctx.get_engine()
+            self.stats_generator = engine.get_component(self.generator_id)
+            if not isinstance(self.stats_generator, Statistics):
+                raise TypeError(
+                    f"statistics generator must implement Statistics type. Got: {type(self.stats_generator)}"
+                )
+            self.stats_generator.initialize(engine.get_all_components(), fl_ctx)
+        except Exception as e:
+            self.log_exception(fl_ctx, f"statistics generator initialize exception: {e}")
+
+    def execute(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
+        client_name = fl_ctx.get_prop(ReservedKey.CLIENT_NAME)
+        self.log_info(fl_ctx, f"Executing task '{task_name}' for client: '{client_name}'")
+        try:
+            result = self.client_exec(task_name, shareable, fl_ctx, abort_signal)
+            if abort_signal.triggered:
+                return make_reply(ReturnCode.TASK_ABORTED)
```

Comment:
the `client_exec` method is blocking,

If we finish it and then found out the abort signal is triggered, maybe we can still send the result back.

Reply:
we might want to check the abort signal before `client_exec` or in between

Reply:
If the `client_exec` is a long/slow method, it might worth it to check the abort signal inside the method.

If it is short, then I agree checking before / or in the end does not matter, we could even ignore the check.